File: license/tools/generate-license.ts
#!/usr/bin/env npx ts-node
/**
 * ライセンスキー発行 CLI ツール
 *
 * 使用方法:
 *   npx ts-node generate-license.ts --product ALL --tier TRIAL
 *   npx ts-node generate-license.ts --product SLIDE --tier PRO --expires 2025-12-31
 *   npx ts-node generate-license.ts --product FORG --tier STD --months 12
 */

import {
  generateLicenseWithExpiry,
  ProductCode,
  LicenseTier,
  PRODUCT_NAMES,
  TIER_NAMES,
  TIERS,
} from '../typescript/index';

// コマンドライン引数解析
function parseArgs(): {
  product: ProductCode;
  tier: LicenseTier;
  expires?: Date;
  months?: number;
  count: number;
} {
  const args = process.argv.slice(2);
  let product: ProductCode = 'ALL';
  let tier: LicenseTier = 'TRIAL';
  let expires: Date | undefined;
  let months: number | undefined;
  let count = 1;

  for (let i = 0; i < args.length; i++) {
    switch (args[i]) {
      case '--product':
      case '-p':
        product = args[++i] as ProductCode;
        break;
      case '--tier':
      case '-t':
        tier = args[++i] as LicenseTier;
        break;
      case '--expires':
      case '-e':
        expires = new Date(args[++i]);
        break;
      case '--months':
      case '-m':
        months = parseInt(args[++i], 10);
        break;
      case '--count':
      case '-n':
        count = parseInt(args[++i], 10);
        break;
      case '--help':
      case '-h':
        printHelp();
        process.exit(0);
    }
  }

  // 月数指定の場合は有効期限を計算
  if (months && !expires) {
    expires = new Date();
    expires.setMonth(expires.getMonth() + months);
  }

  return { product, tier, expires, months, count };
}

function printHelp(): void {
  console.log(`
ライセンスキー発行ツール

使用方法:
  npx ts-node generate-license.ts [オプション]

オプション:
  -p, --product <CODE>   製品コード (SALES, SLIDE, PY, INTV, FORG, ALL)
  -t, --tier <TIER>      ティア (TRIAL, STD, PRO, ENT)
  -e, --expires <DATE>   有効期限 (YYYY-MM-DD形式)
  -m, --months <N>       有効期間（月数）
  -n, --count <N>        発行数
  -h, --help             ヘルプを表示

例:
  # 全製品トライアル（14日間）
  npx ts-node generate-license.ts -p ALL -t TRIAL

  # InsightSlide Professional（12ヶ月）
  npx ts-node generate-license.ts -p SLIDE -t PRO -m 12

  # InsightForguncy Standard（指定日まで）
  npx ts-node generate-license.ts -p FORG -t STD -e 2025-12-31

  # 10個一括発行
  npx ts-node generate-license.ts -p ALL -t TRIAL -n 10

製品コード:
  SALES  - SalesInsight
  SLIDE  - InsightSlide
  PY     - InsightPy
  INTV   - InterviewInsight
  FORG   - InsightForguncy
  ALL    - 全製品バンドル

ティア:
  TRIAL  - トライアル（デフォルト14日）
  STD    - Standard（年間）
  PRO    - Professional（年間）
  ENT    - Enterprise（永久）
`);
}

function formatDate(date: Date | null): string {
  if (!date) return '永久';
  return date.toISOString().split('T')[0];
}

function main(): void {
  const { product, tier, expires, count } = parseArgs();

  // 製品コード検証
  if (!PRODUCT_NAMES[product]) {
    console.error(`❌ 不正な製品コード: ${product}`);
    console.error(`   有効な値: ${Object.keys(PRODUCT_NAMES).join(', ')}`);
    process.exit(1);
  }

  // ティア検証
  if (!TIER_NAMES[tier]) {
    console.error(`❌ 不正なティア: ${tier}`);
    console.error(`   有効な値: ${Object.keys(TIER_NAMES).join(', ')}`);
    process.exit(1);
  }

  console.log('');
  console.log('========================================');
  console.log('  Insight Series ライセンス発行');
  console.log('========================================');
  console.log('');
  console.log(`製品: ${PRODUCT_NAMES[product]} (${product})`);
  console.log(`ティア: ${TIER_NAMES[tier]} (${tier})`);
  console.log(`発行数: ${count}`);
  console.log('');
  console.log('----------------------------------------');

  const licenses: { key: string; expires: string }[] = [];

  for (let i = 0; i < count; i++) {
    const result = generateLicenseWithExpiry({
      productCode: product,
      tier: tier,
      expiresAt: expires,
    });

    const expiresStr = formatDate(result.expiresAt);
    licenses.push({ key: result.licenseKey, expires: expiresStr });

    console.log(`${i + 1}. ${result.licenseKey}`);
    console.log(`   有効期限: ${expiresStr}`);
  }

  console.log('----------------------------------------');
  console.log('');

  // CSV形式で出力
  if (count > 1) {
    console.log('CSV形式:');
    console.log('license_key,expires_at,product,tier');
    licenses.forEach(l => {
      console.log(`${l.key},${l.expires},${product},${tier}`);
    });
    console.log('');
  }

  // JSON形式で出力
  console.log('JSON形式:');
  console.log(JSON.stringify(licenses.map(l => ({
    licenseKey: l.key,
    expiresAt: l.expires,
    product,
    tier,
    productName: PRODUCT_NAMES[product],
    tierName: TIER_NAMES[tier],
  })), null, 2));
}

main();

File: license/tools/generate-license.py
#!/usr/bin/env python3
"""
繝ｩ繧､繧ｻ繝ｳ繧ｹ繧ｭ繝ｼ逋ｺ陦・CLI 繝・・繝ｫ

繧ｭ繝ｼ蠖｢蠑・ PPPP-PLAN-YYMM-HASH-SIG1-SIG2

菴ｿ逕ｨ譁ｹ豕・
  python generate-license.py --product INSS --plan PRO --email user@example.com --expires 2027-01-31
  python generate-license.py --product INSS --trial --email user@example.com
"""

import argparse
import json
import sys
from datetime import datetime, timedelta
from pathlib import Path

# 繝代せ繧定ｿｽ蜉
sys.path.insert(0, str(Path(__file__).parent.parent / 'python'))

from __init__ import (
    generate_license_key,
    generate_trial_key,
    ProductCode,
    Plan,
    PRODUCT_NAMES,
    PLAN_NAMES,
    TRIAL_DAYS,
)


def parse_args():
    parser = argparse.ArgumentParser(
        description='Insight Series 繝ｩ繧､繧ｻ繝ｳ繧ｹ繧ｭ繝ｼ逋ｺ陦後ヤ繝ｼ繝ｫ',
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog='''
萓・
  # InsightOfficeSlide Pro・・027蟷ｴ1譛医∪縺ｧ・・
  python generate-license.py -p INSS --plan PRO -e user@example.com --expires 2027-01-31

  # InsightOfficeSlide Standard・・2繝ｶ譛茨ｼ・
  python generate-license.py -p INSS --plan STD -e user@example.com -m 12

  # InsightPy 繝医Λ繧､繧｢繝ｫ・・4譌･髢難ｼ・
  python generate-license.py -p INPY --trial -e user@example.com

  # InsightCast Pro・・2繝ｶ譛茨ｼ・
  python generate-license.py -p INMV --plan PRO -e user@example.com -m 12

  # InsightBot Pro・・2繝ｶ譛茨ｼ・
  python generate-license.py -p INBT --plan PRO -e user@example.com -m 12

  # InterviewInsight Standard・・2繝ｶ譛茨ｼ・
  python generate-license.py -p IVIN --plan STD -e user@example.com -m 12

陬ｽ蜩√さ繝ｼ繝・
  INSS  - InsightOfficeSlide
  IOSH  - InsightOfficeSheet
  IOSD  - InsightOfficeDoc
  INPY  - InsightPy
  INMV  - InsightCast
  INIG  - InsightImageGen
  INBT  - InsightBot
  INCA  - InsightNoCodeAnalyzer
  IVIN  - InterviewInsight

繝励Λ繝ｳ:
  TRIAL  - 繝医Λ繧､繧｢繝ｫ・・4譌･髢難ｼ・
  STD    - Standard
  PRO    - Professional
'''
    )

    parser.add_argument(
        '-p', '--product',
        type=str,
        required=True,
        choices=['INSS', 'IOSH', 'IOSD', 'INPY', 'INMV', 'INIG', 'INBT', 'INCA', 'IVIN'],
        help='陬ｽ蜩√さ繝ｼ繝・
    )

    parser.add_argument(
        '--plan',
        type=str,
        choices=['STD', 'PRO'],
        help='繝励Λ繝ｳ・・-trial 縺ｨ謗剃ｻ厄ｼ・
    )

    parser.add_argument(
        '--trial',
        action='store_true',
        help='繝医Λ繧､繧｢繝ｫ逋ｺ陦鯉ｼ・4譌･髢難ｼ・
    )

    parser.add_argument(
        '-e', '--email',
        type=str,
        required=True,
        help='繝｡繝ｼ繝ｫ繧｢繝峨Ξ繧ｹ・亥ｿ・茨ｼ・
    )

    parser.add_argument(
        '--expires',
        type=str,
        help='譛牙柑譛滄剞 (YYYY-MM-DD蠖｢蠑・'
    )

    parser.add_argument(
        '-m', '--months',
        type=int,
        help='譛牙柑譛滄俣・域怦謨ｰ・・
    )

    parser.add_argument(
        '--json',
        action='store_true',
        help='JSON蠖｢蠑上・縺ｿ蜃ｺ蜉・
    )

    parser.add_argument(
        '--csv',
        action='store_true',
        help='CSV蠖｢蠑上・縺ｿ蜃ｺ蜉・
    )

    return parser.parse_args()


def format_date(dt):
    if dt is None:
        return '豌ｸ荵・
    return dt.strftime('%Y-%m-%d')


def main():
    args = parse_args()

    # 繝舌Μ繝・・繧ｷ繝ｧ繝ｳ
    if not args.trial and not args.plan:
        print("繧ｨ繝ｩ繝ｼ: --plan 縺ｾ縺溘・ --trial 繧呈欠螳壹＠縺ｦ縺上□縺輔＞", file=sys.stderr)
        sys.exit(1)

    if args.trial and args.plan:
        print("繧ｨ繝ｩ繝ｼ: --plan 縺ｨ --trial 縺ｯ蜷梧凾縺ｫ謖・ｮ壹〒縺阪∪縺帙ｓ", file=sys.stderr)
        sys.exit(1)

    # 陬ｽ蜩√さ繝ｼ繝・
    product_code = ProductCode(args.product)

    # 繝医Λ繧､繧｢繝ｫ縺ｮ蝣ｴ蜷・
    if args.trial:
        license_key = generate_trial_key(product_code, args.email)
        expires_at = datetime.now() + timedelta(days=TRIAL_DAYS)
        plan = Plan.TRIAL
    else:
        plan = Plan(args.plan)

        # 譛牙柑譛滄剞縺ｮ豎ｺ螳・
        if args.expires:
            expires_at = datetime.strptime(args.expires, '%Y-%m-%d')
        elif args.months:
            expires_at = datetime.now()
            new_month = expires_at.month + args.months
            new_year = expires_at.year + (new_month - 1) // 12
            new_month = (new_month - 1) % 12 + 1
            expires_at = datetime(new_year, new_month, min(expires_at.day, 28))
        else:
            # 繝・ヵ繧ｩ繝ｫ繝・ 12繝ｶ譛・
            expires_at = datetime.now()
            new_month = expires_at.month + 12
            new_year = expires_at.year + (new_month - 1) // 12
            new_month = (new_month - 1) % 12 + 1
            expires_at = datetime(new_year, new_month, min(expires_at.day, 28))

        license_key = generate_license_key(product_code, plan, args.email, expires_at)

    result = {
        'licenseKey': license_key,
        'expiresAt': format_date(expires_at),
        'product': args.product,
        'plan': plan.value,
        'email': args.email,
        'productName': PRODUCT_NAMES[product_code],
        'planName': PLAN_NAMES[plan],
    }

    # 蜃ｺ蜉・
    if args.json:
        print(json.dumps(result, indent=2, ensure_ascii=False))
        return

    if args.csv:
        print('license_key,expires_at,product,plan,email')
        print(f"{result['licenseKey']},{result['expiresAt']},{result['product']},{result['plan']},{result['email']}")
        return

    # 騾壼ｸｸ蜃ｺ蜉・
    print('')
    print('========================================')
    print('  Insight Series 繝ｩ繧､繧ｻ繝ｳ繧ｹ逋ｺ陦・)
    print('========================================')
    print('')
    print(f"陬ｽ蜩・       {result['productName']} ({result['product']})")
    print(f"繝励Λ繝ｳ:     {result['planName']} ({result['plan']})")
    print(f"繝｡繝ｼ繝ｫ:     {result['email']}")
    print(f"譛牙柑譛滄剞:   {result['expiresAt']}")
    print('')
    print('----------------------------------------')
    print(f"繝ｩ繧､繧ｻ繝ｳ繧ｹ繧ｭ繝ｼ: {result['licenseKey']}")
    print('----------------------------------------')
    print('')
    print('JSON蠖｢蠑・')
    print(json.dumps(result, indent=2, ensure_ascii=False))


if __name__ == '__main__':
    main()
